model-evaluation/src/main/java/ai/vespa/models/evaluation/OnnxExpressionNode.java

```diff
@@ -0,0 +1,102 @@
+// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package ai.vespa.models.evaluation;
+
+import com.yahoo.searchlib.rankingexpression.Reference;
+import com.yahoo.searchlib.rankingexpression.evaluation.Context;
+import com.yahoo.searchlib.rankingexpression.evaluation.TensorValue;
+import com.yahoo.searchlib.rankingexpression.evaluation.Value;
+import com.yahoo.searchlib.rankingexpression.rule.CompositeNode;
+import com.yahoo.searchlib.rankingexpression.rule.ExpressionNode;
+import com.yahoo.searchlib.rankingexpression.rule.ReferenceNode;
+import com.yahoo.searchlib.rankingexpression.rule.SerializationContext;
+import com.yahoo.tensor.Tensor;
+import com.yahoo.tensor.TensorType;
+import com.yahoo.tensor.evaluation.TypeContext;
+
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * make it possible to evaluate an ONNX model anywhere in the ranking expression tree
+ */
+class OnnxExpressionNode extends CompositeNode {
+    private final OnnxModel model;
+    private final String onnxOutputName;
+    private final TensorType expectedType;
+    private final String outputAs;
+    private final List<String> modelInputs = new ArrayList<>();
+    private final List<ExpressionNode> inputRefs = new ArrayList<>();
+
+    OnnxExpressionNode(OnnxModel model, String onnxOutputName, TensorType expectedType, String outputAs) {
+        this.model = model;
+        this.onnxOutputName = onnxOutputName;
+        this.expectedType = expectedType;
+        this.outputAs = outputAs;
+        for (var input : model.inputSpecs) {
+            modelInputs.add(input.onnxName);
+            var optRef = parseOnnxInput(input.source);
+            if (optRef.isEmpty()) {
+                throw new IllegalArgumentException("Bad input source for ONNX model " + model.name() + ": '" + input + "'");
+            }
+            var ref = optRef.get();
+            inputRefs.add(new ReferenceNode(ref));
+        }
+    }
+
+    static Optional<Reference> parseOnnxInput(String input) {
+        var optRef = Reference.simple(input);
+        if (optRef.isPresent()) {
+            return optRef;
+        }
+        try {
+            var ref = Reference.fromIdentifier(input);
+            return Optional.of(ref);
+        } catch (Exception e) {
+            // fallthrough
```

Comment:
When can this happen?

Reply:
> When can this happen?

only on unexpected configuration; it's there only to make it possible to give a better error message just above.